Add ability to re-run a data script from the the data update script page

Ridhwana opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Note: For brevity I use the abbreviation DUS  when referring to a data update script. 

The PR contains the following features:
- It adds a rerun button on the DUS page to a row that has status of failed. This gives us the ability to rerun a DUS. 
- When the user clicks the rerun button for a particular DUS, it kicks off a job that re runs that DUS. 
- It then starts polling the show DUS endpoint for an updated status.
- Once it received that updated status, it uses JS magic to update the DOM to reflect that. 
- It also handles and displays errors appropriately on the page.

NOTE:
I want to note that when we clicked on the re run button multiple times in development, I was getting the error `:error => "NoMethodError: undefined method `new' for nil:NilClass"` for  the line of code `script.file_class.new.run` in app/workers/data_update_worker.rb:25. 

Rhymes and I paired on this for sometime and we deduced that we think is happening because of the way that the development environment loads up the files. If we set the config in development to mimic the production environment by changing the following values to true, we no longer see the error happening. 
```
  # In the development environment your application's code is reloaded on
  # every request. This slows down response time but is perfect for development
  # since you don't have to restart the web server when you make code changes.
  config.cache_classes = false

  # Do not eager load code on boot.
  config.eager_load = false
```
We will have to double check this in production. This should be low risk as the dev team will be using this feature mostly at first.

## Related Tickets & Documents
https://github.com/forem/internalEngineering/issues/276
An older issue: https://github.com/forem/forem/issues/6175

## QA Instructions, Screenshots, Recordings

<img width="995" alt="Screenshot 2021-02-01 at 13 53 12" src="https://user-images.githubusercontent.com/2786819/106455497-f4354000-6494-11eb-81cd-a8dc2b51995c.png">

**Watch video on how to replicate an DUS failure on dev and then click re-run to make it succeed:**
https://www.loom.com/share/33b31be15a2b458ab3b44a4e5ad55992

**To summarize the video:**

Setting up the env for failure first:
- Go to http://localhost:3000/admin/data_update_scripts
- Get the filename from row 1, and find it in your editor
- Update the file to have some syntax error that will cause the script to fail. 

- Now delete the data script that you've updated from the database by finding the id using `DataUpdateScript.find(some-id).destroy`
- Rerun the data update scripts with `rails data_updates:run` so that the failure can take effect.

Setting the development env to mimic prod:
- Go to config.development.rb 
- Change to `config.eager_load = true` and   `config.cache_classes = true`

Going to http://localhost:3000/admin/data_update_scripts you should see a failure.
Now stop your server, fix the file that you tempered with.
Going back to http://localhost:3000/admin/data_update_scripts and click Rerun. When you click re run you will see it chnage to succeeded.

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

- [x] Yes 
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed as yet

## [optional] Are there any post deployment tasks we need to perform?
None 

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/xBqg5gAf1xINizpek6/giphy.gif)

Ridhwana commented on app/views/admin/data_update_scripts/index.html.erb:
```diff
@@ -7,23 +7,66 @@
+      .then(json => {
+
+        let updatedScript = json.response;
+        // escape out the <>
```
To fix: Will remove this comment once I fix this issue. The html in the error is being rendered as html instead of being shown as a string.

rhymes commented on app/controllers/admin/data_update_scripts_controller.rb:
```diff
@@ -3,7 +3,16 @@ class DataUpdateScriptsController < Admin::ApplicationController
+      @data_update_scripts = DataUpdateScript.order(run_at: :desc)
     end
+
+    def run_data_update_scripts
+
+      data_update_script_id = params[:id]
+      response = DataUpdateScript.force_run(data_update_script_id)
+
+      render json: { response: response }
+    end
```
This should obviously take into account the possibility that the id doesn't exist when `force_run` raises a 404

rhymes commented on app/views/admin/data_update_scripts/index.html.erb:
```diff
@@ -7,23 +7,66 @@
   </table>
 </main>
+
+<script>
+  function forceRun(id) {
```
can we move this into a Stimulus controller? I think inline JS should be avoided as much as possible as it also makes modularization difficult and it's ignored by ESLint as far as i know

rhymes commented on config/routes.rb:
```diff
@@ -50,6 +50,7 @@
         mount flipper_ui, at: "feature_flags"
 
         resources :data_update_scripts, only: [:index]
+        post "/run_data_update_scripts", to: "data_update_scripts#run_data_update_scripts"
```
as the URL only run a single script it should probably be singular. Even though @citizen428 would say here to use RESTy routes on a different controller I think we can add the extra route here but use `resources` properly using [member routes](https://guides.rubyonrails.org/routing.html#adding-member-routes). Something like: `POST /data_update_scripts/{id}/force-run` or something like that

rhymes commented on app/models/data_update_script.rb:
```diff
@@ -25,6 +25,40 @@ def scripts_to_run?
       false
     end
 
+    def force_run(id)
+      script = DataUpdateScript.find(id)
+      log_status(script)
+      run_script(script)
+
+      script
+    end
+
+    def run_script(script)
+      require script.file_path
+      script.file_class.new.run
+
+      script.mark_as_finished!
+      log_status(script)
+    rescue StandardError => e
+      script.mark_as_failed!(e)
+      log_status(script)
+
+      Honeybadger.notify(e, context: { script_id: script.id })
+    end
+
+    def log_status(script)
+      status = script.status
+      file_name = script.file_name
+
+      logger_destination = status.to_sym == :failed ? :error : :info
+      Rails.logger.public_send(
+        logger_destination,
+        "time=#{Time.current.rfc3339}, script=#{file_name}, status=#{status}",
+      )
+
+      DatadogStatsClient.increment("data_update_scripts.status", tags: ["status:#{status}", "script_name:#{file_name}"])
+    end
```
As you mentioned privately, these contain business logic that should probably be splint in a service that's called in the controller and the `DataUpdateWorker`

rhymes commented:
> Should I perhaps only allow the failed scripts to be re-run instead of all scripts?

I'd double check with @maestromac here but I think we do not want scripts that have passed successfully to be retried. Especially if they are old or have been commented out :)

maestromac commented:
@Ridhwana I think that if you feel the need to rerun a succeeded script, it means that a particular script is flawed and warrants a replacement/adjustment. I think `re-run` should only be reserved for failed scripts. It's also a surface area for mistakes; people could accidentally rerun a succeeded script.

On a different note, this can be ignored and addressed in the future: once in a while, we would get a stop script. Those could use a `force stop` button, which then would switch to a `re-run` after the script is properly halted.

Ridhwana commented:
> @Ridhwana I think that if you feel the need to rerun a succeeded script, it means that a particular script is flawed and warrants a replacement/adjustment. I think `re-run` should only be reserved for failed scripts. It's also a surface area for mistakes; people could accidentally rerun a succeeded script.

I thought so, this sounds good, thanks @maestromac.
 
> On a different note, this can be ignored and addressed in the future: once in a while, we would get a stop script. Those could use a `force stop` button, which then would switch to a `re-run` after the script is properly halted.
👍🏽

Ridhwana commented on app/controllers/admin/data_update_scripts_controller.rb:
```diff
@@ -3,7 +3,13 @@ class DataUpdateScriptsController < Admin::ApplicationController
+    end
+
+    def force_run
+      response = DataUpdateScript.force_run(params[:id])
```
@citizen428 hmm this makes sense. Do you have a suggestion on the best method to tackle this, I was thinking maybe a sidekiq worker?

citizen428 replied:
Yes, the re-run should be async.

To update the UI, maybe we could add a `show` action to this controller that returns the status of a single DUS as JSON. Then when you trigger the re-run in the Stimulus controller, you schedule a function with `setTimeOut` or something that periodically checks this endpoint: if the script has finished, update the UI and `clearTimeout`, if it's still in progress reschedule.

But maybe all of this is too complicated and we just schedule the script in Sidekiq, do a full page refresh and put a message in the flash like "Re-run scheduled, this may take a bit"

citizen428 commented on app/controllers/admin/data_update_scripts_controller.rb:
```diff
@@ -9,6 +9,8 @@ def index
     def show
       response = DataUpdateScript.find(params[:id])
       render json: { response: response }
+    rescue StandardError => e
```
**nitpick (non-blocking):** We catch `StandardError`, but return a status of `not_found`. How about we rescue the narrowest exception possible?

```suggestion
    rescue ActiveRecord::RecordNotFound=> e
```

Ridhwana commented on config/routes.rb:
```diff
@@ -50,7 +50,8 @@
         mount flipper_ui, at: "feature_flags"
 
-        resources :data_update_scripts, only: [:index]
+        resources :data_update_scripts, only: %i[index show]
```
We add a show route that we poll for an updated status once the job is run

rhymes commented on app/controllers/admin/data_update_scripts_controller.rb:
```diff
@@ -3,7 +3,20 @@ class DataUpdateScriptsController < Admin::ApplicationController
+    def force_run
+      DataUpdateWorker.perform_async(params[:id])
+    rescue StandardError => e
+      render json: { error: "#{e.class}: #{e.message}" }, status: :unprocessable_entity
```
I'm not sure there's a scenario when `StandardError` can be triggered here as you're executing an async script. The error inside the script will be triggered by the script, not by the controller:

```ruby
pry(main)> DataUpdateWorker.perform_async("abcdef")
=> "2fe553e19238c6d3ec7471c9"
```

you can pass whatever, it's Sidekiq that's going to fail when it deserializes an invalid ID

Ridhwana replied:
@rhymes Do you think that the rescue block in this case in unnecessary then?

rhymes replied:
@Ridhwana yeah. there's no scenario when `.perform_async` fails. Even if you don't pass required articles to it it won't fail. It will fail when the job is picked up by Sidekiq

rhymes commented on app/javascript/admin/controllers/data_update_script_controller.js:
```diff
@@ -0,0 +1,114 @@
+
+export default class DataUpdateScriptController extends Controller {
+
+  forceRun(event) {
```
Would that have worked by giving the same target to every script and adding a `data-value` with the `id` to fetch inside Stimulus? 

Stimulus, since 2.0, supports [multiple HTML elements with the same target](https://stimulus.hotwire.dev/reference/targets):

> Plural | this.[name]Targets | An array of all matching targets in scope

Anyway, probably not worth revisiting for now

Ridhwana replied:
Yes possibly, I must have been looking at slightly older docs. I'm going to add a note for possibly my cooldown to refactor out this code and try it this way.

rhymes commented on app/javascript/admin/controllers/data_update_script_controller.js:
```diff
@@ -0,0 +1,114 @@
+          clearInterval(pollForStatus);
+        }
+      });
+      if ( counter > 20 ) {
```
Not likely but the downside with this mechanism is that it will have 1 timer for each retried script, instead of checking them as a batch. Just an idea for future improvements

Ridhwana replied:
Thanks Rhymes, this makes sense too, and I've noted it for a future improvement.

rhymes commented on app/models/data_update_script.rb:
```diff
@@ -48,7 +48,7 @@ def mark_as_run!
 
   def mark_as_finished!
-    update!(finished_at: Time.current, status: :succeeded)
+    update!(finished_at: Time.current, status: :succeeded, error: nil)
```
One thing that might want to add in the future is: when did it fail?

Ridhwana replied:
Do you mean the time that it failed? wouldn't that just be the run_at time?

rhymes replied:
Oh sorry, we have `finished_at`, it was right in front of my eyes :D

Ridhwana replied:
Lol mine too 😅

rhymes reviewed (commented):
I put a `sleep(30)` in my script and it worked well. Only one cosmetic detail:

`This may take some time. Please refresh the page to check for the status.`

is shown as a red error, maybe an `info` notice with better? IIRC they are yellow in bootstrap

Ridhwana commented:
> I put a sleep(30) in my script and it worked well. Only one cosmetic detail:This may take some time. Please refresh the page to check for the status. is shown as a red error, maybe an info notice with better? IIRC they are yellow in bootstrap

Awesome, thanks so much for the review :) Makes sense, I'll sort this out sometime later today.

Ridhwana commented:
@rhymes I've done the following:

1. I've removed the rescue block in the controller
2. I've changed the banner styles depending on what we're showing.
An error will still look the same:
<img width="995" alt="Screenshot 2021-02-01 at 13 53 12" src="https://user-images.githubusercontent.com/2786819/106492176-50618980-64c0-11eb-8353-21145bd1c6da.png">

The please refresh message is now an info banner that has a blue background. 
<img width="1010" alt="Screenshot 2021-02-01 at 18 51 30" src="https://user-images.githubusercontent.com/2786819/106492198-56576a80-64c0-11eb-907a-20898521e7f7.png">

Ridhwana commented on app/javascript/admin/__tests__/controllers/data_update_script_controller.test.js:
```diff
@@ -0,0 +1,92 @@
+import { Application } from 'stimulus';
```
I see a potential to possibly add more Stimulus tests, however this seems to cover at least the basis. In some downtime when I refactor the Stimulus to use data-target as per Rhymes suggestion, I may add some more.

mstruve reviewed (approved):
The frontend code is over my head but I pulled down the branch and it works great!!! 

Two thoughts
1. `Data Script with 248 may take some time. Please refresh the page to check for the status.` Maybe use the file name instead of the ID for the flash message if possible
2. Should we do a confirm box with the button?

Ridhwana commented:
> Data Script with 248 may take some time. Please refresh the page to check for the status. Maybe use the file name instead of the ID for the flash message if possible
@mstruve sounds good, I'll update the filename in a follow up PR.

> Should we do a confirm box with the button?
I'm not sure that we need this right now. Only the failed DUSs show a re-run button and at first only the Forem team (with already restricted access) will most likely be using it. I think it would make sense to see how it's being used in production and then add these additional features. We should probably also be showing the users the content of the files instead of them having to switch back and then confirm that they want to re-run the script.

Ridhwana commented:
PR https://github.com/forem/forem/pull/12529 should fix unrelated failing test.

rhymes reviewed (approved):
LGTM!
